### ISSUE TITLE
[REVIEW]: QComms QKD Software Toolkit

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#1119]
**Submitting author:** @richardcollins (<a href="http://orcid.org/0000-0002-6238-990X">Richard Collins</a>)
**Repository:** <a href="https://gitlab.com/QComms/cqptoolkit" target ="_blank">https://gitlab.com/QComms/cqptoolkit</a>
**Version:** 0.3
**Editor:** @brainstorm
**Reviewers:** @arturgs
**Archive:** <a href="https://doi.org/10.5281/zenodo.3247364" target="_blank">10.5281/zenodo.3247364</a>

## Status

[![status](http://joss.theoj.org/papers/f652455b21fbc39b784e86c765a64fe8/status.svg)](http://joss.theoj.org/papers/f652455b21fbc39b784e86c765a64fe8)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/f652455b21fbc39b784e86c765a64fe8"><img src="http://joss.theoj.org/papers/f652455b21fbc39b784e86c765a64fe8/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/f652455b21fbc39b784e86c765a64fe8/status.svg)](http://joss.theoj.org/papers/f652455b21fbc39b784e86c765a64fe8)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://gitlab.com/QComms/cqptoolkit" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@arturgs, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.theoj.org/about#reviewer_guidelines. Any questions/concerns please let @brainstorm know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://gitlab.com/QComms/cqptoolkit">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (0.3)?
- [x] **Authorship:** Has the submitting author (@richardcollins) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://gitlab.com/QComms/cqptoolkit">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (0.3)?
- [ ] **Authorship:** Has the submitting author (@richardcollins) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @ejdanderson it looks like you're currently assigned as the reviewer for this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01119/joss.01119/10.21105.joss.01119.pdf)

**Comment by ejdanderson** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List Whedon's capabilities
@whedon commands

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

# Compile the paper
@whedon generate pdf

```

**Comment by brainstorm** [comment]
@ejdanderson You might want to have a look at https://joss.readthedocs.io/en/latest/review_criteria.html for some guidance on how to tick the boxes above, let me know if you have any doubts about the process. When in doubt, you can check other issues in this same issue tracker and see how different software reviews went, thanks again for your efforts!

**Comment by brainstorm** [comment]
:wave: @ejdanderson Let me know if you have any doubts on how to proceed :)

**Comment by ejdanderson** [comment]
Great thanks! I'll have time this weekend to review.

**Comment by ejdanderson** [comment]
@brainstorm  There appears to be plenty of documentation, tests, and example programs to draw from for the software. Unfortunately, I imagine most people will be installing this via docker, which I was unable to do following the instructions (see Installation below). 

### Authorship:

- Djeylan Aktas  - It is unclear of Atkas’ contribution to the software. Are they an academic advisor?

### Installation: 

I have attempted to install this via the docker installation and was presented an error (even post typo fix) see [issue 6](https://gitlab.com/QComms/cqptoolkit/issues/6).

### Community guidelines: 

It would be worthwhile to explicitly state how contributions, issues, and support can be made via the README. Either a separate contribution.md file ([example](https://github.com/MarionLegrandLogrami/stacomiR/blob/master/CONTRIBUTING.md)) or a few lines in the README itself is recommended. I found the Coding.md file to be very informative, thank you for including this.

###  Software paper

- @brainstorm Can you comment on the requirement of having a Summary heading? The introduction serves as a valid summary, but likely needs to be changed to "Summary"

- Either expand the approach heading or remove it.

**Comment by richardcollins** [comment]
Hello Roman, sorry for the delay, I've been on holiday. I will take a look at the comments asap.

Thankyou @ejdanderson for reviewing the project.

________________________________

**Comment by richardcollins** [comment]
@ejdanderson I've changed the heading in the paper to Summary.
@brainstorm  wrt the contributions, Djeylan and others have provided details on the algorithms and the technical aspects of QKD but havn't written code in this project - do I need to cite them in a different way?

**Comment by brainstorm** [comment]
@richardcollins According to the [JOSS reviewer guidelines](https://joss.readthedocs.io/en/latest/review_criteria.html#authorship) you, as the author, take the responsibility for authorship claims, so thanks for stating it here, that should suffice AFAICT.

@ejdanderson Feel free to carry on with the remaining items whenever you can.

Cheers!

**Comment by ejdanderson** [comment]
Thanks, I'll take a look this week.

**Comment by brainstorm** [comment]
@ejdanderson Let me know if you have any doubts regarding any of the review points, happy to help!

**Comment by brainstorm** [comment]
👋 @ejdanderson, sorry to bother you with this, if it's tricky to test/install that's also a valid feedback for the author :)

**Comment by richardcollins** [comment]
@ejdanderson, Sorry to bug you about this but are you able to spend some time on this? - I'm happy to answer questions. I need to report my publications in a progress report and this has been noted as stalled.

**Comment by labarba** [comment]
At this point, we're beset by a delinquent reviewer, and one possible course of action is to find a second reviewer who is willing to fast-track this submission.

@brainstorm — try to find someone willing to make such a contribution to JOSS. You'll need to manually edit the first post in this thread to add a new checklist for the reviewer. If they have already reviewed for JOSS, they will have permissions to tick off the items. If not, we'll have to add them here https://github.com/openjournals/joss-reviews/settings/collaborationa

**Comment by brainstorm** [comment]
I'm contacting two additional quantum computing researchers through a common contact, hopefully they'll be able and willing to review and fast-track this paper. Thanks for your patience @richardcollins and @labarba for the additional editorial pointers, this is the first case of many papers that a reviewer went missing for me.

**Comment by brainstorm** [comment]
Hello @arturgs, @brunojulia recommended you as a possible reviewer for this quantum key distribution paper, would you be interested in reviewing it? The process is fairly straightforward and semi-automated via GitHub:

https://joss.readthedocs.io/en/latest/reviewer_guidelines.html

Let me know if you are interested!

**Comment by brainstorm** [comment]
Hello @SwamyDev, I peeked at your https://github.com/SwamyDev/q_network QKD framework and thought that you could assist JOSS (Journal of Open Source Software) on reviewing this software:

https://gitlab.com/QComms/cqptoolkit

The review process is held via this issue and here on Github, please let us know if you are interested in reviewing this software publication.

**Comment by danielskatz** [comment]
👋 @brainstorm - what's happening here?  It looks like this has been waiting for 2 new reviewer for about 6 weeks?

**Comment by brainstorm** [comment]
@whedon assign @arturgs as reviewer

**Comment by whedon** [comment]
OK, the reviewer is @arturgs

**Comment by danielskatz** [comment]
👋 @arfon, can you add @arturgs to the group so he can review?

I'm going to create a new set of review criteria/checkmarks for him above, in the first comment in this issue.

**Comment by arfon** [comment]
> 👋 @arfon, can you add @arturgs to the group so he can review?

Done. @arturgs you'll need to accept the invite at https://github.com/openjournals/joss-reviews/invitations before you'll be able to update the checklist above.

**Comment by arfon** [comment]
:wave: @arturgs - could you give us an update on the status of your review?

**Comment by arturgs** [comment]
@arfon - completed

**Comment by brainstorm** [comment]
@arturgs There's a few unchecked checkboxes above, but I'll take your word for it, thanks a lot for your review!

@richardcollins Could you please generate a Zenodo DOI so that we can archive and finally publish? Thanks for your patience, this has been a tough case to close! :)

**Comment by richardcollins** [comment]
Thankyou @arturgs for reviewing this.
@brainstorm I've created an upload in [zenodo](https://zenodo.org/record/3247364) with a DOI of 10.5281/zenodo.3247364
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.3247364.svg)](https://doi.org/10.5281/zenodo.3247364)

**Comment by arturgs** [comment]
Hi all, this was my first review. 
Just a comment: once finished, how to announce that the review is done?

**Comment by brainstorm** [comment]
@arturgs No worries, you did well. The only detail was making sure you went through all the checkboxes above and then just chiming in the comment that you have gone through those ;)

**Comment by brainstorm** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Attempting to check references...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.1080/09500340008244057 is OK
- 10.1364/OFC.2018.M2A.6 is OK
- 10.1109/JLT.2016.2646921 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by brainstorm** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01119/joss.01119/10.21105.joss.01119.pdf)

**Comment by richardcollins** [comment]
looks good to me, thanks

**Comment by brainstorm** [comment]
@whedon set 10.5281/zenodo.3247364 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3247364" target="_blank">10.5281/zenodo.3247364</a> is the archive.

**Comment by brainstorm** [comment]
@arfon Looks good to me, ready to ship, thanks everyone!

**Comment by arfon** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.1080/09500340008244057 is OK
- 10.1364/OFC.2018.M2A.6 is OK
- 10.1109/JLT.2016.2646921 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/766

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/766, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by arfon** [comment]
@richardcollins - there looks to be an empty section in your paper without any content:

```markdown
# Approach

````

If this is the actual heading, then the headings following it need to use smaller headings (e.g.  `## Components` vs `# Components`, `### Statistics` vs `## Statistics`).

Also, this section is superfluous as we link to the software from the paper metadata in the left-hand margin:

```markdown
# Access to the code
The code can be found [on GitLab](https://gitlab.com/QComms/cqptoolkit.git)
```

**Comment by richardcollins** [comment]
@arfon Sorry about that - not sure how I missed that.
I've updated the paper.md to remove those sections. Do I need to update the zenodo version with this new version?

**Comment by arfon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01119/joss.01119/10.21105.joss.01119.pdf)

**Comment by arfon** [comment]
@richardcollins thanks, I we don't need a Zenodo update for paper changes. A couple of other suggested changes:

- Unless you really want to have it there, I'd suggest removing your title from your author name.
- Could you please say a little bit more in the introduction about `QKD`, i.e. _what_ it is (perhaps define the term?) and what it is commonly used for? JOSS papers are designed for a non-specialist audience and so it's important that the introduction provides this context.

**Comment by richardcollins** [comment]
How about adding this before "The system demonstrates..."?
Quantum key distribution (QKD) uses properties of quantum states such as the [no-cloning theorem](https://www.quantiki.org/wiki/no-cloning-theorem) to share randomness between different locations, without the need to physically transport the key material. This randomness is known to be secure because the probability of an eavesdropper can be calculated. This means that [symmetric encryption](https://www.ssl2buy.com/wiki/symmetric-vs-asymmetric-encryption-what-are-differences) instead of public key can be used, which can prevent attacks from [quantum computers](https://www.cryptomathic.com/news-events/blog/quantum-computing-and-its-impact-on-cryptography). Symmetric keys are only as secure as the medium they are transported over. Unlike QKD, manually exchanged key is expensive (e.g. physically driving somewhere with a flash drive) and has to be used sparingly.  Making guaranteed secure key available as a resource makes symmetric keys more viable.

**Comment by arfon** [comment]
👍 yes that looks like a good addition. Thanks @richardcollins

**Comment by richardcollins** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01119/joss.01119/10.21105.joss.01119.pdf)

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01119/joss.01119/10.21105.joss.01119.pdf)

**Comment by arfon** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.1080/09500340008244057 is OK
- 10.1364/OFC.2018.M2A.6 is OK
- 10.1109/JLT.2016.2646921 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/770

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/770, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by arfon** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1140696959797354496) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/771
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.01119](https://doi.org/10.21105/joss.01119)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by arfon** [comment]
@arturgs - many thanks for your review here and to @brainstorm for editing this submission ✨ 

@richardcollins - your paper is now accepted into JOSS :zap::rocket::boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](http://joss.theoj.org/papers/10.21105/joss.01119/status.svg)](https://doi.org/10.21105/joss.01119)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.01119">
  <img src="http://joss.theoj.org/papers/10.21105/joss.01119/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: http://joss.theoj.org/papers/10.21105/joss.01119/status.svg
   :target: https://doi.org/10.21105/joss.01119
```
This is how it will look in your documentation:

[![DOI](http://joss.theoj.org/papers/10.21105/joss.01119/status.svg)](https://doi.org/10.21105/joss.01119)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: http://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.salsalabs.org/donate-to-joss

**Comment by richardcollins** [comment]
Thankyou so much @arfon
